### PR TITLE
[AutoDiff] Added support for 'withoutDerivative(at:)'.

### PR DESCRIPTION
**Pull request by eaplatanios**
@rxwei following from our discussion in tensorflow/swift-apis#150.

### CODE REVIEW

**Comment by eaplatanios** [comment]
One question that remains is, given these new helpers do we want to keep the `Differentiable.withoutDerivative()` helper? Removing it would likely result in a more consistent API with respect to how we denote code segments without derivatives.

**Comment by rxwei** [comment]
> One question that remains is, given these new helpers do we want to keep the `Differentiable.withoutDerivative()` helper? Removing it would likely result in a more consistent API with respect to how we denote code segments without derivatives.

No, please feel free to remove it.

**Comment by eaplatanios** [comment]
Sounds good. I'll do that but I need to know how to replace [this](https://github.com/apple/swift/blob/def2be969bb8212f6cfc388698a4a58d4c4283b5/lib/SILOptimizer/Mandatory/Differentiation.cpp#L4277). Is there something along the lines of `fixItInsertAround(sourceLoc, { x in "withoutDerivative(at: \(x))")`?

**Comment by rxwei** [comment]
There is a `fixItInsert` (before) and a `fixItInsertAfter`. A call to each will work.

File: stdlib/public/core/AutoDiff.swift [hunk 1]
```diff
@@ -147,6 +147,16 @@ public extension Differentiable {
   func withoutDerivative() -> Self { return self }
 }
 
+/// Acts as an identity function. When used in a context where `x` is being
```
Comment (rxwei) [inline stdlib/public/core/AutoDiff.swift]:
```suggestion
/// Returns `x` like an identity function. When used in a context where `x` is being
```

**Comment by eaplatanios** [comment]
Thanks! I made these edits and also changed the relevant tests. I cannot currently test locally so could you please run the CI tests so I check if the error reporting test modifications are correct?

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 2]
```diff
@@ -4274,7 +4274,8 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
       if (sourceLoc.isValid()) {
         getContext()
             .diagnose(sourceLoc, diag::autodiff_nonvaried_result_fixit)
-            .fixItInsertAfter(sourceLoc, ".withoutDerivative()");
+            .fixItInsert(sourceLoc, "withoutDerivative(at:")
+            .fixItInsertAfter(sourceLoc, ")");
```
Comment (rxwei) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]: You might need to lookup an expression first and get its source range, from which you can get its start loc and end loc.

Comment (eaplatanios) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp]: I just pushed a fix. Would that work?

Comment (rxwei) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp]: Looks like it might work. There may be many tricky edge cases. Did you update tests?

**Review by rxwei — APPROVED** [review]

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

File: lib/SILOptimizer/Mandatory/Differentiation.cpp (hunk 2 above)
Comment (eaplatanios) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp]: Yes I can see that. I hope that by working on incremental fixes and features in the compiler codebase I may gain a basic understanding of what's going on overall and what the common/useful patterns are.

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by eaplatanios** [comment]
The CI errors are due to swift-apis. I'll make a corresponding PR there and try to test locally a bit later today.

**Comment by eaplatanios** [comment]
@rxwei Tests seem to pass locally now. Could you please start the CI tests?

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please clean test tensorflow

**Pull request merged**